Sources/NIO/ControlMessage.swift

```diff
@@ -0,0 +1,251 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2020 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+import CNIODarwin
+#elseif os(Linux) || os(FreeBSD) || os(Android)
+import CNIOLinux
+#endif
+
+/// Representation of a `cmsghdr` and associated data.
+/// Unsafe as captures pointers and must not escape the scope where those pointers are valid.
+struct UnsafeControlMessage {
+    var level: CInt
+    var type: CInt
+    var data: UnsafeRawBufferPointer?
+}
+
+/// Collection representation of `cmsghdr` structures and associated data from `recvmsg`
+/// Unsafe as captures pointers held in msghdr structure which must not escape scope of validity.
+struct UnsafeControlMessageCollection {
+    private var messageHeader: msghdr
+
+    init(messageHeader: msghdr) {
+        self.messageHeader = messageHeader
+    }
+}
+
+// Add the `Collection` functionality to UnsafeControlMessageCollection.
+extension UnsafeControlMessageCollection: Collection {
+    typealias Element = UnsafeControlMessage
+    
+    struct Index: Equatable, Comparable {
+        fileprivate var cmsgPointer: UnsafeMutablePointer<cmsghdr>?
+        
+        static func < (lhs: UnsafeControlMessageCollection.Index,
+                       rhs: UnsafeControlMessageCollection.Index) -> Bool {
+            // nil is high, as that's the end of the collection.
+            switch (lhs.cmsgPointer, rhs.cmsgPointer) {
+            case (.some(let lhs), .some(let rhs)):
+                return lhs < rhs
+            case (.some, .none):
+                return true
+            case (.none, .some), (.none, .none):
+                return false
+            }
+        }
+        
+        fileprivate init(cmsgPointer: UnsafeMutablePointer<cmsghdr>?) {
+            self.cmsgPointer = cmsgPointer
+        }
+    }
+    
+    var startIndex: Index {
+        var messageHeader = self.messageHeader
+        return withUnsafePointer(to: &messageHeader) { messageHeaderPtr in
+            let firstCMsg = Posix.cmsgFirstHeader(inside: messageHeaderPtr)
+            return Index(cmsgPointer: firstCMsg)
+        }
+    }
+    
+    var endIndex: Index { return Index(cmsgPointer: nil) }
+    
+    func index(after: Index) -> Index {
+        precondition(after.cmsgPointer != nil)
```

Comment:
This precondition is unnecessary: we force-unwrap this pointer before its first use anyway, so the correctness of the code isn't enforced here. If you wanted to match the message below you can move the force-unwrap to this line instead, if that helps.